replication.go

```diff
@@ -144,8 +158,91 @@ func (r *ReplicationState) sendRequestToNode(start uint64, end uint64, node Node
 	}
 	msg := &Message{ReplicationRequest: request}
 
-	r.lastSequenceRequested = end
-	r.comm.SendMessage(msg, node)
+	task := r.createReplicationTimeoutTask(start, end, nodes, index)
+
+	r.timeoutHandler.AddTask(task)
+
+	r.comm.SendMessage(msg, nodes[index])
+}
+
+func (r *ReplicationState) createReplicationTimeoutTask(start, end uint64, nodes []NodeID, index int) *TimeoutTask {
+	taskFunc := func() {
+		r.sendRequestToNode(start, end, nodes, (index+1)%len(nodes))
+	}
+	timeoutTask := &TimeoutTask{
+		Start:    start,
+		End:      end,
+		NodeID:   nodes[index],
+		TaskID:   getTimeoutID(start, end),
+		Task:     taskFunc,
+		Deadline: r.timeoutHandler.GetTime().Add(DefaultReplicationRequestTimeout),
+	}
+
+	return timeoutTask
+}
+
+// receivedReplicationResponse notifies the task handler a response was received. If the response
+// was incomplete(meaning our timeout expected more seqs), then we will create a new timeout
+// for the missing sequences and send the request to a different node.
+func (r *ReplicationState) receivedReplicationResponse(data []QuorumRound, node NodeID) {
+	seqs := make([]uint64, 0, len(data))
+
+	for _, qr := range data {
+		seqs = append(seqs, qr.GetSequence())
+	}
+
+	slices.Sort(seqs)
+
+	task := r.timeoutHandler.FindTask(node, seqs)
+	if task == nil {
+		return
+	}
+	r.timeoutHandler.RemoveTask(node, task.TaskID)
+
+	// we found the timeout, now make sure all seqs were returned
+	missing := findMissingNumbersInRange(task.Start, task.End, seqs)
+	if len(missing) == 0 {
+		return
+	}
+
+	// if not all sequences were returned, create new timeouts
+	nodes := r.highestSequenceObserved.signers.Remove(r.id)
+	lenNodes := len(nodes)
+
+	start := missing[0]
```

Comment:
```
	start := missing[0]
	end := missing[0]
	for i, missingSeq := range missing[1:] {
		if missingSeq != missing[i]+1 {
			index := i % lenNodes
			newTask := r.createReplicationTimeoutTask(start, end, nodes, index)
			r.timeoutHandler.AddTask(newTask)
			start = missingSeq
		}

		end = missingSeq
```

Is it possible to abstract out the creation of the sections into a function (let's denote it `f() []struct{start uint64; end uint64}` and add a UT for it, and then just do:

```
for _, startEndPair := range f() { 
		start := startEnd.start
		end := startEnd.end
                 newTask := r.createReplicationTimeoutTask(start, end, nodes, index)
}
```

?